packages/components/src/ui/create-styles/components/theme-provider/theme-provider.js

```diff
@@ -0,0 +1,170 @@
+ * @property {Record<string, string>} [darkHighContrastTheme={}] Custom dark & high contrast theme properties.
+ */
+
+/**
+ * The ThemeProvider for the Style system. This ThemeProvider uses Emotion's
+ * ThemeProvider as a foundation, but enhances it with features provided by
+ * the Style system, such as dark mode, high contrast mode, etc...
+ *
+ * An important feature this ThemeProvider accounts for is the ability to render
+ * styles either globally (at the document/html level) or scoped.
+ *
+ * @example
+ * ```jsx
+ * <ThemeProvider isGlobal theme={{...}}>
+ * <Button>...</Button>
+ * </ThemeProvider>
+ * ```
+ * @param {ThemeProviderProps} props Props for the ThemeProvider.
+ * @param {import('react').Ref<any>} forwardedRef
+ * @return {JSX.Element} Children content wrapped with the <ThemeProvider />.
+ */
+function ThemeProvider(
+	{
+		children,
+		compiler,
+		className,
+		isGlobal = false,
+		globalStyles,
+		isDark = false,
+		isColorBlind = false,
+		isReducedMotion = false,
+		isHighContrast = false,
+		theme = {},
+		darkTheme = {},
+		highContrastTheme = {},
+		darkHighContrastTheme = {},
+		...props
+	},
+	forwardedRef
+) {
+	const { css, cx, injectGlobal } = compiler;
+	/**
+	 * Hydrates global styles (via injectGlobal). This is necessary as there may
+	 * be a chance that <ThemeProvider /> renders before any other (styled)
+	 * component. Injecting global styles early in this manner allows for
+	 * the initial render of theme styles (which also uses injectGlobal)
+	 * to be sequenced correctly.
+	 */
+	useHydrateGlobalStyles( { injectGlobal, globalStyles } );
+
+	const nodeRef = useRef();
+	const defaultStyles = useThemeStyles( {
+		injectGlobal,
+		isGlobal,
+		theme,
+		selector: ':root',
+	} );
+	const darkStyles = useThemeStyles( {
+		injectGlobal,
+		isGlobal,
+		theme: darkTheme,
+		selector: repeat( DARK_MODE_ATTR, MODE_SPECIFICITY_COMPOUND_LEVEL ),
+	} );
+	const highContrastStyles = useThemeStyles( {
+		injectGlobal,
+		isGlobal,
+		theme: highContrastTheme,
+		selector: repeat(
+			HIGH_CONTRAST_MODE_MODE_ATTR,
+			MODE_SPECIFICITY_COMPOUND_LEVEL
+		),
+	} );
+	const darkHighContrastStyles = useThemeStyles( {
+		injectGlobal,
+		isGlobal,
+		theme: darkHighContrastTheme,
+		selector: repeat(
+			DARK_HIGH_CONTRAST_MODE_MODE_ATTR,
+			MODE_SPECIFICITY_COMPOUND_LEVEL
+		),
+	} );
+
+	useColorBlindMode( { isColorBlind, isGlobal, ref: nodeRef } );
+	useDarkMode( { isDark, isGlobal, ref: nodeRef } );
+	useHighContrastMode( { isGlobal, isHighContrast, ref: nodeRef } );
+	useReducedMotionMode( { isGlobal, isReducedMotion, ref: nodeRef } );
+
+	const contextState = useThemeProviderContextBridge( {
+		isDark,
+		isReducedMotion,
+		isColorBlind,
+		isHighContrast,
+	} );
+
+	const modeHtmlAttrs = useThemeProviderModeHtmlAttributes( contextState );
+
+	const classes = cx(
+		className,
+		css`
+			${ defaultStyles }
+			${ darkStyles }
+			${ highContrastStyles }
+			${ darkHighContrastStyles }
+		`
+	);
+
+	return (
+		<div
```

Comment:
Why do we need a "div" for the theme provider?

Reply:
In other words, in my mind, the theme provider just sets a "context" and let the components consume the data from that theme. so I'd like to understand more why we need a div. We tend to be as pure as possible and try to kill divs as much as we can in the editor and Gutenberg related screens.